Add proxy functionality

DannyAziz opened the pull request:
This will allow you to pass in a proxy to mask your requests.

In the CLI you pass in the raw url for the proxy

In code you can pass in the dict for the proxy:
```
{"http": "http://user:pass@domain:port", "https": "https://user:pass@domain:port"}
```

jdepoix commented:
Hi @DannyAziz,

thanks a lot for your contribution! 

Could you please elaborate a bit more on what the use case for this is?

jdepoix commented on README.md:
````diff
@@ -84,6 +84,23 @@ If you would prefer to write it into a file or pipe it into another application,
+
+CLI:
+```
+youtube_transcript_api <first_video_id> <second_video_id> http://user:pass@domain:port
````
```suggestion
youtube_transcript_api <first_video_id> <second_video_id> --proxy http://user:pass@domain:port
```

jdepoix replied:
The CLI doesn't allow for specifying individual proxies for HTTP and HTTPS and will implicitly use the HTTP proxy on HTTPS. I don't know if this will be a problem for your use case, but maybe you should consider using a `--http-proxy` and `--https-proxy` param.

jdepoix commented on youtube_transcript_api/_api.py:
```diff
@@ -37,7 +37,7 @@ def __init__(self, video_id):
 
     @staticmethod
-    def get_transcripts(video_ids, languages=None, continue_after_error=False):
+    def get_transcripts(video_ids, languages=None, continue_after_error=False, proxy=None):
```
Please update Docstrings of all public methods accordingly.

jdepoix commented on youtube_transcript_api/_api.py:
```diff
@@ -70,7 +70,7 @@ def get_transcripts(video_ids, languages=None, continue_after_error=False):
 
     @staticmethod
-    def get_transcript(video_id, languages=None):
+    def get_transcript(video_id, languages=None, proxy=None):
```
Please update Docstrings of all public methods accordingly.

jdepoix commented on youtube_transcript_api/_api.py:
```diff
@@ -100,12 +101,19 @@ class _TranscriptFetcher():
+    def __init__(self, video_id, languages, proxy):
         self.video_id = video_id
         self.languages = languages
+        if isinstance(proxy, str):
```
I'd prefer the input of this to be consistently typed. The handling of different inputs should be done while parsing the arguments.

DannyAziz commented:
Hi @jdepoix, I'm currently using the project to get transcriptions on mass and even though I'm doing it in a friendly manner (batched requests with timeouts and delayed requests) to not ping Youtube's servers too hard - I'm a bit concerned about all these requests coming from one IP address.

I prefer this project over `youtube-dl` which already has the proxy functionality, so I'd like to keep using it

I'll get to work on your reviews

jdepoix commented on youtube_transcript_api/_api.py:
```diff
@@ -83,10 +85,13 @@ def get_transcript(video_id, languages=None):
         :return: a list of dictionaries containing the 'text', 'start' and 'duration' keys
         :rtype: [{'text': str, 'start': float, 'end': float}]
+        :param proxy: a dictionary mapping of http and https proxies to be used for the network requests
+        :rtype {'http': http_proxy, 'https': https_proxy}
```
```suggestion
        :rtype {'http': str, 'https': str}
```

jdepoix replied:
The docstring and README could also reference http://docs.python-requests.org/en/master/user/advanced/#proxies. As this describes the possible inputs in more detail.

jdepoix commented on youtube_transcript_api/_api.py:
```diff
@@ -100,12 +105,16 @@ class _TranscriptFetcher():
+    def __init__(self, video_id, languages, proxy):
         self.video_id = video_id
         self.languages = languages
+        self.proxy = proxy
```
thinking about it, `proxies` would probably be the more appropriate name, as it is also used for requests.

jdepoix commented:
> Hi @jdepoix, I'm currently using the project to get transcriptions on mass and even though I'm doing it in a friendly manner (batched requests with timeouts and delayed requests) to not ping Youtube's servers too hard - I'm a bit concerned about all these requests coming from one IP address.
> 
> I prefer this project over `youtube-dl` which already has the proxy functionality, so I'd like to keep using it
> 
> I'll get to work on your reviews

Thank you for sharing your usecase. To be honest I don't really share your concerns, however I could see proxies being useful to bypass geo restrictions.

jdepoix commented on youtube_transcript_api/__main__.py:
```diff
@@ -39,18 +39,31 @@ def parse_args(args):
 
     parsed_args = parse_args(sys.argv[1:])
+
+    proxies = {"http": parsed_args.http_proxy or "", "https": parsed_args.https_proxy or ""}
```
Sorry, I just noticed this. But proxies has to be `None` if neither `parsed_args.http_proxy` nor `parsed_args.https_proxy` are set. Otherwise the `proxies` dict with both empty strings will be used.

jdepoix replied:
```suggestion
    if parsed_args.http_proxy or parsed_args.https_proxy:
        proxies = {"http": parsed_args.http_proxy or "", "https": parsed_args.https_proxy or ""}
    else:
        proxies = None
```

jdepoix commented on youtube_transcript_api/__main__.py:
```diff
@@ -39,18 +39,31 @@ def parse_args(args):
-
+    parser.add_argument(
+        '--http-proxy', dest='http_proxy',
+        default=None, metavar='URL',
```
if you use `default=''` here, you won't have to do the conditional assignment in `proxies = {"http": parsed_args.http_proxy or "", "https": parsed_args.https_proxy or ""}`

jdepoix commented:
Hey @DannyAziz,

to make contributions like yours more effortless I just added a full test suite and set up a CI pipeline in #11. It would be great if you could merge master into this branch and add tests for your code!

DannyAziz commented:
Hey, @jdepoix thanks for this! I've refactored my code a little and added tests but I'm wondering if you have any ideas on mocking the proxies for the API tests - I looked through the docs for `httpretty` but couldn't find anything useful

jdepoix commented:
Hey @DannyAziz,
I don't think that it is necessary to test if the http requests which is done by the requests library actually uses the proxy. It is a functionality of the request library, therefore it is their responsibility to test it and make sure that it works. Therefore, if you just test if `request.get(...)` is called with the right params (depending on the proxy being set or not), you can assume that `request.get(...)` will do the right things!

jdepoix commented on youtube_transcript_api/test/test_cli.py:
```diff
@@ -23,6 +23,31 @@ def test_argument_parsing(self):
+        self.assertEqual(parsed_args.video_ids, ['v1', 'v2'])
+        self.assertEqual(parsed_args.json, True)
+        self.assertEqual(parsed_args.languages, ['de', 'en'])
+        self.assertEqual(parsed_args.http_proxy, 'http://user:pass@domain:port')
```
also make sure that the arguments which are not explicitly set default to the expected value (`https_proxy`)

jdepoix commented on youtube_transcript_api/test/test_cli.py:
```diff
@@ -23,6 +23,31 @@ def test_argument_parsing(self):
+        self.assertEqual(parsed_args.video_ids, ['v1', 'v2'])
+        self.assertEqual(parsed_args.json, True)
+        self.assertEqual(parsed_args.languages, ['de', 'en'])
+        self.assertEqual(parsed_args.https_proxy, 'https://user:pass@domain:port')
```
same here

jdepoix commented on youtube_transcript_api/test/test_cli.py:
```diff
@@ -50,14 +75,26 @@ def test_argument_parsing__languages(self):
             languages=['de', 'en'],
-            continue_after_error=True
+            continue_after_error=True,
+            proxies={"http": "", "https": ""}
```
I guess if no proxy options are set you would want `proxies=None`. Otherwise the `if proxies: ...` calls which happen later on will always be true and therefore `requests.get(..., proxies={...})` will always be called even though no proxy options are set.

jdepoix commented on youtube_transcript_api/test/test_cli.py:
```diff
@@ -50,14 +77,26 @@ def test_argument_parsing__languages(self):
+        )._parse_args()
+        self.assertEqual(parsed_args.https_proxy, 'https://user:pass@domain:port')
+
     def test_run(self):
```
You'll also have to run a version of this test with proxies being set, to get full coverage

DannyAziz replied:
Thanks for letting me know!

DannyAziz replied:
So I just moved the coverage a tiny bit, perhaps I misunderstood what you meant by test with proxies being set?

jdepoix replied:
Yeah, you might have misunderstood me here :)

During the build coveralls checks if all branches of the code are covered by the tests which are run. You can see the results here: https://coveralls.io/builds/22302114/source?filename=youtube_transcript_api%2F_cli.py#L19

As you can see the part of the `run()` method which is only called when a proxy is set, is not executed during any of the tests. Since the tests you added only test `_parse_args()` but not `run()` they won't cover this line. To do so you'll need to add the following test:

```python
def test_run__proxies(self):
    YouTubeTranscriptApi.get_transcripts = MagicMock(return_value=([], []))
    YouTubeTranscriptCli('v1 v2 --languages de en --http-proxy http://user:pass@domain:port --https-proxy https://user:pass@domain:port'.split()).run()

    YouTubeTranscriptApi.get_transcripts.assert_called_once_with(
        ['v1', 'v2'],
        languages=['de', 'en'],
        continue_after_error=True,
        proxies={'http': 'http://user:pass@domain:port', 'https': 'https://user:pass@domain:port'}
    )
```

jdepoix commented on youtube_transcript_api/test/test_api.py:
```diff
@@ -5,7 +5,7 @@
 import httpretty
 
-from youtube_transcript_api._api import YouTubeTranscriptApi
+from youtube_transcript_api._api import YouTubeTranscriptApi, _TranscriptFetcher
```
I can't see `_TranscriptFetcher` being used. If so, please remove the import.

jdepoix commented on youtube_transcript_api/test/test_cli.py:
```diff
@@ -50,6 +50,15 @@ def test_argument_parsing(self):
         self.assertEqual(parsed_args.http_proxy, '')
 
+        parsed_args = YouTubeTranscriptCli(
+            'v1 v2 --languages de en --json'.split()
```
These arguments are equivalent to the second testcase within `test_argument_parsing`. However I think the expected values for `parsed_args.http_proxy` and `parsed_args.https_proxy` should be added to all testcases.

jdepoix commented:
@DannyAziz I will now merge this and release it in v0.1.4.

Thank you for you contribution and taking your time for the proposed changes! 👍

jdepoix commented:
This feature has now been released as part of v0.1.4!
